File: py/safe_ci_dag_runner/reservation.py
"""Stateful CPU-core RESERVATION ledger — collision-free core allocation.

``pick_least_busy_free_cores`` (in :mod:`safe_ci_dag_runner.cgroup`) prevents a
single run from OVER-using cores: it samples ``/proc/stat`` and returns the K
least-busy of this process's allowed set. That heuristic cannot prevent
COLLISION — two benchmarks launched at the same instant sample the same idle
cores and both take them, so their tracer/tracee threads land on the same CPUs
and the "pinned" measurement is contaminated by an unrelated run. Sampling has
no memory of who already holds what.

This module adds the missing state: a durable, ``flock``-serialized ledger that
maps each reserved core to its holder. Every acquire runs under an exclusive
lock, so it observes every prior live reservation and picks from the FREE-and-
UNHELD set only. The caller NEVER picks a core itself.

Design invariants (each has a test in ``tests/test_reservation.py``):

  * DISJOINT under concurrency — two acquires that overlap in time get disjoint
    core sets. The exclusive ``flock`` is held across the whole
    sweep→pick→record critical section (including the ``/proc/stat`` sample), so
    a concurrent acquire blocks, then sees the first's cores as HELD and
    excludes them. Serializing acquires is the price of disjointness and is
    correct for benchmark setup (a handful of concurrent requests, not a hot
    path).
  * RELEASE on completion — :meth:`Reservation.release`, the
    :func:`reserve_cores` context manager (``finally``), and an ``atexit`` hook
    all free the cores. Normal exit and exceptions both release.
  * DEAD-HOLDER RECLAIM — a crashed holder cannot release. Every acquire first
    sweeps the ledger and drops records whose holder process is no longer live,
    so a leaked reservation is reclaimed by the next acquire instead of
    permanently subtracting cores (the same failure class as leaked cgroup
    scopes that held cores because systemd only reaps an EMPTY scope). Liveness
    is fingerprinted by (pid, /proc starttime) so a recycled PID is NOT mistaken
    for the original holder.
"""

from __future__ import annotations

import atexit
import fcntl
import json
import os
import tempfile
import time
from collections.abc import Collection, Iterable
from dataclasses import dataclass, field
from pathlib import Path
from types import TracebackType
from typing import IO

from safe_ci_dag_runner.cgroup import pick_least_busy_free_cores


class InsufficientCoresError(RuntimeError):
    """Raised when fewer than the requested K cores are free-and-unheld.

    The allocator refuses to hand out a colliding core: it is better to fail
    loudly than to return a set that overlaps a live reservation."""


def _default_ledger_path() -> Path:
    """Ledger location, shared by every process on the host that reserves cores.

    Prefers ``$XDG_RUNTIME_DIR`` (a per-user tmpfs reaped on logout), falling
    back to a uid-scoped ``/tmp`` dir. The whole point is a SINGLE file all
    concurrent runners contend on, so the path must not vary by CWD."""
    env = os.environ.get("SAFE_CI_CORE_LEDGER")
    if env:
        return Path(env)
    base = os.environ.get("XDG_RUNTIME_DIR")
    if base and os.path.isdir(base):
        root = Path(base) / "safe-ci-dag-runner"
    else:
        root = Path(tempfile.gettempdir()) / f"safe-ci-dag-runner-{os.getuid()}"
    return root / "core-reservations.json"


def _proc_starttime(pid: int) -> int | None:
    """The holder's start time (clock ticks since boot, ``/proc/<pid>/stat``
    field 22). Combined with the PID it fingerprints a process across PID reuse:
    a recycled PID has a different starttime, so a dead holder is never mistaken
    for a live one. Returns ``None`` if the process is gone."""
    try:
        data = Path(f"/proc/{pid}/stat").read_text()
    except (FileNotFoundError, ProcessLookupError, PermissionError):
        return None
    # comm (field 2) is parenthesized and may contain spaces/parens; everything
    # after the LAST ')' is space-separated. starttime is field 22 → after comm
    # the fields start at field 3 (index 0), so starttime is index 22-3 = 19.
    rparen = data.rfind(")")
    if rparen < 0:
        return None
    rest = data[rparen + 2 :].split()
    if len(rest) <= 19:
        return None
    try:
        return int(rest[19])
    except ValueError:
        return None


def _holder_alive(pid: int, starttime: int | None) -> bool:
    """True iff a process with this exact (pid, starttime) fingerprint is live.

    A bare ``kill(pid, 0)`` success is not enough — the PID may have been
    recycled by an unrelated process. Requiring the recorded starttime to match
    defeats that."""
    cur = _proc_starttime(pid)
    if cur is None:
        return False
    if starttime is None:
        # Legacy record without a fingerprint: fall back to bare liveness.
        return True
    return cur == starttime


@dataclass
class _Record:
    pid: int
    starttime: int | None
    cores: list[int]
    tag: str
    ts: float

    @classmethod
    def from_json(cls, d: dict[str, object]) -> "_Record":
        st = d.get("starttime")
        cores_raw = d.get("cores", [])
        cores = [int(str(c)) for c in cores_raw] if isinstance(cores_raw, list) else []
        return cls(
            pid=int(str(d.get("pid", 0))),
            starttime=(int(str(st)) if st is not None else None),
            cores=cores,
            tag=str(d.get("tag", "")),
            ts=float(str(d.get("ts", 0.0))),
        )

    def to_json(self) -> dict[str, object]:
        return {
            "pid": self.pid,
            "starttime": self.starttime,
            "cores": self.cores,
            "tag": self.tag,
            "ts": self.ts,
        }


class _LedgerLock:
    """Exclusive ``flock`` over the ledger's critical section.

    The lock lives on a sibling ``.lock`` file (never the JSON itself, which is
    replaced by atomic rename). Held across sweep→pick→record so the whole
    allocation is serialized and disjoint."""

    def __init__(self, ledger: Path):
        self._lock_path = ledger.with_suffix(ledger.suffix + ".lock")
        self._lock_path.parent.mkdir(parents=True, exist_ok=True)
        self._fh: IO[str] | None = None

    def __enter__(self) -> "_LedgerLock":
        self._fh = open(self._lock_path, "w")
        fcntl.flock(self._fh.fileno(), fcntl.LOCK_EX)
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._fh is None:
            return
        fcntl.flock(self._fh.fileno(), fcntl.LOCK_UN)
        self._fh.close()
        self._fh = None


def _load(ledger: Path) -> list[_Record]:
    try:
        raw: object = json.loads(ledger.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return []
    if not isinstance(raw, dict):
        return []
    recs = raw.get("reservations", [])
    if not isinstance(recs, list):
        return []
    out: list[_Record] = []
    for r in recs:
        if isinstance(r, dict):
            out.append(_Record.from_json(r))
    return out


def _store(ledger: Path, records: Iterable[_Record]) -> None:
    """Atomically replace the ledger (write temp in the same dir, then rename)."""
    ledger.parent.mkdir(parents=True, exist_ok=True)
    payload = {"reservations": [r.to_json() for r in records]}
    fd, tmp = tempfile.mkstemp(dir=str(ledger.parent), suffix=".tmp")
    try:
        with os.fdopen(fd, "w") as fh:
            json.dump(payload, fh)
        os.replace(tmp, ledger)
    except BaseException:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


def _sweep(records: list[_Record]) -> tuple[list[_Record], list[_Record]]:
    """Split records into (live, reclaimed-dead) by holder liveness."""
    live: list[_Record] = []
    dead: list[_Record] = []
    for r in records:
        (live if _holder_alive(r.pid, r.starttime) else dead).append(r)
    return live, dead


@dataclass
class Reservation:
    """A held set of cores. Release exactly once (idempotent)."""

    cores: list[int]
    pid: int
    starttime: int | None
    tag: str
    ledger: Path
    _released: bool = field(default=False, repr=False)

    def release(self) -> None:
        """Free this reservation's cores. Idempotent and crash-safe: even if
        this is never called, the next acquire reclaims the record."""
        if self._released:
            return
        with _LedgerLock(self.ledger):
            records = _load(self.ledger)
            kept = [
                r
                for r in records
                if not (r.pid == self.pid and r.starttime == self.starttime and r.tag == self.tag)
            ]
            _store(self.ledger, kept)
        self._released = True

    def __enter__(self) -> "Reservation":
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        self.release()


def acquire(
    k: int,
    *,
    tag: str = "",
    sample_s: float = 0.3,
    ledger: Path | None = None,
    exclude: Collection[int] = (),
) -> Reservation:
    """Reserve K disjoint cores for this process, collision-free.

    Under the exclusive ledger lock: sweep dead holders, compute the HELD set
    from live reservations, pick K least-busy cores from the free-and-unheld
    set, record them against this (pid, starttime), and register an ``atexit``
    release. Raises :class:`InsufficientCoresError` if fewer than K cores are
    available — never returns an overlapping set.

    ``exclude`` is unioned into the held set (cores the caller wants avoided for
    reasons outside the ledger, e.g. a reserved housekeeping CPU)."""
    if k < 1:
        raise ValueError(f"k must be >= 1, got {k}")
    path = ledger or _default_ledger_path()
    pid = os.getpid()
    starttime = _proc_starttime(pid)

    with _LedgerLock(path):
        records = _load(path)
        live, _dead = _sweep(records)
        held: set[int] = set(int(c) for c in exclude)
        for r in live:
            held.update(r.cores)
        cores = pick_least_busy_free_cores(k, sample_s=sample_s, exclude=held)
        if len(cores) < k:
            # Persist the sweep even on failure, so leaked cores are reclaimed.
            _store(path, live)
            raise InsufficientCoresError(
                f"requested {k} core(s) but only {len(cores)} free-and-unheld "
                f"(held by {len(live)} live reservation(s): "
                f"{sorted(held)}; allowed set may also be smaller)"
            )
        rec = _Record(pid=pid, starttime=starttime, cores=cores, tag=tag, ts=time.time())
        _store(path, live + [rec])

    reservation = Reservation(
        cores=cores, pid=pid, starttime=starttime, tag=tag, ledger=path
    )
    atexit.register(reservation.release)
    return reservation


def reclaim_dead(ledger: Path | None = None) -> list[dict[str, object]]:
    """Sweep the ledger and drop records whose holder is dead. Returns the
    reclaimed records (as JSON dicts) for logging/tests. Safe to call anytime;
    ``acquire`` also sweeps, so this is mostly for maintenance/inspection."""
    path = ledger or _default_ledger_path()
    with _LedgerLock(path):
        records = _load(path)
        live, dead = _sweep(records)
        if dead:
            _store(path, live)
    return [r.to_json() for r in dead]


def held_cores(ledger: Path | None = None) -> list[int]:
    """The currently-held (live) cores across all reservations. Sweeps dead
    holders first so the answer reflects reality, not leaked records."""
    path = ledger or _default_ledger_path()
    with _LedgerLock(path):
        records = _load(path)
        live, dead = _sweep(records)
        if dead:
            _store(path, live)
        out: set[int] = set()
        for r in live:
            out.update(r.cores)
    return sorted(out)


class reserve_cores:
    """Context manager: acquire on enter, release on exit (normal or exception).

        with reserve_cores(1, tag="ptrace-bench") as cores:
            # HARD-pin the whole tree via a transient scope's cgroup cpuset.
            # (sched_setaffinity is ESCAPABLE — a child can widen its own mask;
            #  mutation-verified 2026-08-04. Use AllowedCPUs, not affinity.)
            subprocess.run(["systemd-run", "--user", "--scope", "--collect",
                            f"-pAllowedCPUs={','.join(map(str, cores))}",
                            "--", *benchmark_cmd])

    The `cpuset-alloc run --cores K -- CMD` CLI
    (:mod:`safe_ci_dag_runner.cpuset_allocator`) wraps exactly this."""

    def __init__(
        self,
        k: int,
        *,
        tag: str = "",
        sample_s: float = 0.3,
        ledger: Path | None = None,
        exclude: Collection[int] = (),
    ):
        self._k = k
        self._tag = tag
        self._sample_s = sample_s
        self._ledger = ledger
        self._exclude = exclude
        self._reservation: Reservation | None = None

    def __enter__(self) -> list[int]:
        self._reservation = acquire(
            self._k,
            tag=self._tag,
            sample_s=self._sample_s,
            ledger=self._ledger,
            exclude=self._exclude,
        )
        return self._reservation.cores

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._reservation is not None:
            self._reservation.release()

File: rs/safe-ci-dag-runner/src/reservation.rs
//! Durable, collision-free CPU-core reservations shared by local processes.

// Durable, cross-process CPU-core reservations.
//
// The ledger is shared with the Python distribution: both implementations lock the same sibling
// `.lock` file, read/write the same JSON schema, fingerprint holders by `(pid, /proc starttime)`,
// reclaim dead holders, and select only from free allowed CPUs.  A reservation is released on
// `Drop`; a crash is recovered by the next ledger operation.

use std::collections::HashSet;
use std::fs::{self, File, OpenOptions};
use std::io::{Read, Write};
use std::os::fd::AsRawFd;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::{Map, Number, Value};

use crate::cgroup::pick_least_busy_free_cores_excluding;

/// Environment variable overriding the host-wide reservation ledger.
pub const LEDGER_ENV: &str = "SAFE_CI_CORE_LEDGER";

#[derive(Clone, Debug, PartialEq)]
struct Record {
    pid: u32,
    starttime: Option<u64>,
    cores: Vec<usize>,
    tag: String,
    ts: f64,
}

impl Record {
    fn from_value(value: &Value) -> Option<Self> {
        let obj = value.as_object()?;
        let pid = value_u64(obj.get("pid")?)? as u32;
        let starttime = match obj.get("starttime") {
            Some(Value::Null) | None => None,
            Some(v) => value_u64(v),
        };
        let cores = obj
            .get("cores")?
            .as_array()?
            .iter()
            .filter_map(|v| value_u64(v).map(|n| n as usize))
            .collect();
        let tag = value_text(obj.get("tag")).unwrap_or_default();
        let ts = match obj.get("ts") {
            Some(Value::Number(n)) => n.as_f64().unwrap_or(0.0),
            Some(Value::String(s)) => s.parse().unwrap_or(0.0),
            _ => 0.0,
        };
        Some(Self {
            pid,
            starttime,
            cores,
            tag,
            ts,
        })
    }

    fn to_value(&self) -> Value {
        let mut obj = Map::new();
        obj.insert("pid".into(), Value::from(self.pid));
        obj.insert(
            "starttime".into(),
            self.starttime.map(Value::from).unwrap_or(Value::Null),
        );
        obj.insert(
            "cores".into(),
            Value::Array(self.cores.iter().copied().map(Value::from).collect()),
        );
        obj.insert("tag".into(), Value::String(self.tag.clone()));
        obj.insert(
            "ts".into(),
            Number::from_f64(self.ts)
                .map(Value::Number)
                .unwrap_or_else(|| Value::from(0)),
        );
        Value::Object(obj)
    }
}

fn value_u64(value: &Value) -> Option<u64> {
    match value {
        Value::Number(n) => n.as_u64(),
        Value::String(s) => s.parse().ok(),
        _ => None,
    }
}

fn value_text(value: Option<&Value>) -> Option<String> {
    match value {
        Some(Value::String(s)) => Some(s.clone()),
        Some(Value::Number(n)) => Some(n.to_string()),
        Some(Value::Bool(v)) => Some(v.to_string()),
        _ => None,
    }
}

/// Resolve the host-wide reservation-ledger path from the environment or runtime directory.
pub fn default_ledger_path() -> PathBuf {
    if let Some(path) = std::env::var_os(LEDGER_ENV) {
        return PathBuf::from(path);
    }
    if let Some(runtime) = std::env::var_os("XDG_RUNTIME_DIR") {
        let runtime = PathBuf::from(runtime);
        if runtime.is_dir() {
            return runtime
                .join("safe-ci-dag-runner")
                .join("core-reservations.json");
        }
    }
    // SAFETY: geteuid has no preconditions and does not access memory.
    let uid = unsafe { libc::geteuid() };
    std::env::temp_dir()
        .join(format!("safe-ci-dag-runner-{uid}"))
        .join("core-reservations.json")
}

fn proc_starttime(pid: u32) -> Option<u64> {
    let text = fs::read_to_string(format!("/proc/{pid}/stat")).ok()?;
    let rparen = text.rfind(')')?;
    text.get(rparen + 2..)?
        .split_whitespace()
        .nth(19)?
        .parse()
        .ok()
}

fn holder_alive(record: &Record) -> bool {
    match proc_starttime(record.pid) {
        None => false,
        Some(_) if record.starttime.is_none() => true,
        Some(current) => record.starttime == Some(current),
    }
}

struct LedgerLock {
    file: File,
}

impl LedgerLock {
    fn acquire(ledger: &Path) -> Result<Self, String> {
        let lock_path = PathBuf::from(format!("{}.lock", ledger.display()));
        if let Some(parent) = lock_path.parent() {
            fs::create_dir_all(parent).map_err(|e| format!("create ledger directory: {e}"))?;
        }
        let file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .truncate(false)
            .open(&lock_path)
            .map_err(|e| format!("open ledger lock {}: {e}", lock_path.display()))?;
        // SAFETY: flock operates on this live file descriptor and does not retain a pointer.
        if unsafe { libc::flock(file.as_raw_fd(), libc::LOCK_EX) } != 0 {
            return Err(format!(
                "lock reservation ledger {}: {}",
                lock_path.display(),
                std::io::Error::last_os_error()
            ));
        }
        Ok(Self { file })
    }
}

impl Drop for LedgerLock {
    fn drop(&mut self) {
        // SAFETY: the descriptor remains valid until `file` is dropped after this method.
        let _ = unsafe { libc::flock(self.file.as_raw_fd(), libc::LOCK_UN) };
    }
}

fn load(path: &Path) -> Vec<Record> {
    let mut text = String::new();
    if File::open(path)
        .and_then(|mut f| f.read_to_string(&mut text))
        .is_err()
    {
        return Vec::new();
    }
    let Ok(Value::Object(root)) = serde_json::from_str::<Value>(&text) else {
        return Vec::new();
    };
    root.get("reservations")
        .and_then(Value::as_array)
        .map(|items| items.iter().filter_map(Record::from_value).collect())
        .unwrap_or_default()
}

fn store(path: &Path, records: &[Record]) -> Result<(), String> {
    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    fs::create_dir_all(parent).map_err(|e| format!("create ledger directory: {e}"))?;
    let payload = Value::Object(Map::from_iter([(
        "reservations".to_string(),
        Value::Array(records.iter().map(Record::to_value).collect()),
    )]));
    let bytes = serde_json::to_vec(&payload).map_err(|e| format!("encode ledger: {e}"))?;
    let mut created: Option<(PathBuf, File)> = None;
    for attempt in 0..100u32 {
        let temp = parent.join(format!(
            ".core-reservations.{}.{}.tmp",
            std::process::id(),
            attempt
        ));
        match OpenOptions::new().write(true).create_new(true).open(&temp) {
            Ok(file) => {
                created = Some((temp, file));
                break;
            }
            Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => continue,
            Err(e) => return Err(format!("create temporary ledger: {e}")),
        }
    }
    let (temp, mut file) =
        created.ok_or_else(|| "could not allocate temporary ledger".to_string())?;
    let result = (|| {
        file.write_all(&bytes)
            .map_err(|e| format!("write temporary ledger: {e}"))?;
        file.sync_all()
            .map_err(|e| format!("sync temporary ledger: {e}"))?;
        drop(file);
        fs::rename(&temp, path).map_err(|e| format!("replace reservation ledger: {e}"))
    })();
    if result.is_err() {
        let _ = fs::remove_file(&temp);
    }
    result
}

fn sweep(records: Vec<Record>) -> (Vec<Record>, Vec<Record>) {
    records.into_iter().partition(holder_alive)
}

/// A collision-free core reservation.  Dropping it releases its ledger record.
#[derive(Debug)]
pub struct Reservation {
    /// Exact reserved CPU IDs.
    pub cores: Vec<usize>,
    pid: u32,
    starttime: Option<u64>,
    tag: String,
    ledger: PathBuf,
    released: bool,
}

impl Reservation {
    /// Release this reservation.  Idempotent.
    pub fn release(&mut self) -> Result<(), String> {
        if self.released {
            return Ok(());
        }
        let _lock = LedgerLock::acquire(&self.ledger)?;
        let kept: Vec<Record> = load(&self.ledger)
            .into_iter()
            .filter(|r| {
                !(r.pid == self.pid
                    && r.starttime == self.starttime
                    && r.tag == self.tag
                    && r.cores == self.cores)
            })
            .collect();
        store(&self.ledger, &kept)?;
        self.released = true;
        Ok(())
    }

    /// Path of the ledger holding this reservation.
    pub fn ledger(&self) -> &Path {
        &self.ledger
    }
}

impl Drop for Reservation {
    fn drop(&mut self) {
        let _ = self.release();
    }
}

/// Reserve `k` disjoint, least-busy cores for this process.
pub fn acquire(
    k: i64,
    tag: &str,
    sample_s: f64,
    ledger: Option<&Path>,
    exclude: &HashSet<usize>,
    max_irq_rate: Option<f64>,
) -> Result<Reservation, String> {
    if k < 1 {
        return Err(format!("k must be >= 1, got {k}"));
    }
    if !sample_s.is_finite() || sample_s < 0.0 {
        return Err(format!("sample_s must be finite and >= 0, got {sample_s}"));
    }
    let path = ledger
        .map(Path::to_path_buf)
        .unwrap_or_else(default_ledger_path);
    let pid = std::process::id();
    let starttime = proc_starttime(pid);
    let _lock = LedgerLock::acquire(&path)?;
    let (live, _) = sweep(load(&path));
    let mut held = exclude.clone();
    for record in &live {
        held.extend(record.cores.iter().copied());
    }
    let cores = pick_least_busy_free_cores_excluding(k, sample_s, &held, max_irq_rate);
    if cores.len() < k as usize {
        store(&path, &live)?;
        return Err(format!(
            "requested {k} core(s) but only {} free-and-unheld (held by {} live reservation(s): {:?}; allowed set may also be smaller)",
            cores.len(),
            live.len(),
            {
                let mut values: Vec<usize> = held.into_iter().collect();
                values.sort_unstable();
                values
            }
        ));
    }
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0);
    let mut updated = live;
    updated.push(Record {
        pid,
        starttime,
        cores: cores.clone(),
        tag: tag.to_string(),
        ts,
    });
    store(&path, &updated)?;
    Ok(Reservation {
        cores,
        pid,
        starttime,
        tag: tag.to_string(),
        ledger: path,
        released: false,
    })
}

/// Reclaim and return dead-holder records.
pub fn reclaim_dead(ledger: Option<&Path>) -> Result<Vec<Value>, String> {
    let path = ledger
        .map(Path::to_path_buf)
        .unwrap_or_else(default_ledger_path);
    let _lock = LedgerLock::acquire(&path)?;
    let (live, dead) = sweep(load(&path));
    if !dead.is_empty() {
        store(&path, &live)?;
    }
    Ok(dead.iter().map(Record::to_value).collect())
}

/// Return all currently held cores after reclaiming dead holders.
pub fn held_cores(ledger: Option<&Path>) -> Result<Vec<usize>, String> {
    let path = ledger
        .map(Path::to_path_buf)
        .unwrap_or_else(default_ledger_path);
    let _lock = LedgerLock::acquire(&path)?;
    let (live, dead) = sweep(load(&path));
    if !dead.is_empty() {
        store(&path, &live)?;
    }
    let mut cores: Vec<usize> = live
        .iter()
        .flat_map(|record| record.cores.iter().copied())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();
    cores.sort_unstable();
    Ok(cores)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_ledger(name: &str) -> PathBuf {
        std::env::temp_dir().join(format!(
            "safe-ci-rs-reservation-{}-{name}.json",
            std::process::id()
        ))
    }

    #[test]
    fn acquire_release_and_shared_schema() {
        let path = temp_ledger("release");
        let _ = fs::remove_file(&path);
        let mut reservation =
            acquire(1, "unit", 0.001, Some(&path), &HashSet::new(), None).unwrap();
        assert_eq!(held_cores(Some(&path)).unwrap(), reservation.cores);
        let parsed: Value = serde_json::from_str(&fs::read_to_string(&path).unwrap()).unwrap();
        assert!(parsed["reservations"][0]["starttime"].is_number());
        reservation.release().unwrap();
        assert!(held_cores(Some(&path)).unwrap().is_empty());
        let _ = fs::remove_file(&path);
        let _ = fs::remove_file(format!("{}.lock", path.display()));
    }

    #[test]
    fn acquire_rejects_invalid_sample_windows() {
        let path = temp_ledger("sample-window");
        for sample_s in [-1.0, f64::NAN, f64::INFINITY] {
            let error = acquire(1, "unit", sample_s, Some(&path), &HashSet::new(), None)
                .expect_err("invalid sample window must be rejected");
            assert!(error.contains("sample_s must be finite and >= 0"));
        }
    }

    #[test]
    fn dead_holder_is_reclaimed() {
        let path = temp_ledger("dead");
        let dead = Record {
            pid: u32::MAX,
            starttime: Some(1),
            cores: vec![0],
            tag: "dead".into(),
            ts: 0.0,
        };
        store(&path, &[dead]).unwrap();
        let reclaimed = reclaim_dead(Some(&path)).unwrap();
        assert_eq!(reclaimed.len(), 1);
        assert!(held_cores(Some(&path)).unwrap().is_empty());
        let _ = fs::remove_file(&path);
        let _ = fs::remove_file(format!("{}.lock", path.display()));
    }

    #[test]
    fn releasing_one_same_tag_reservation_keeps_the_other() {
        if crate::cgroup::pick_least_busy_free_cores(2, 0.0).len() < 2 {
            return;
        }
        let path = temp_ledger("same-tag");
        let _ = fs::remove_file(&path);
        let mut first = acquire(1, "same", 0.001, Some(&path), &HashSet::new(), None).unwrap();
        let mut second = acquire(1, "same", 0.001, Some(&path), &HashSet::new(), None).unwrap();
        assert_ne!(first.cores, second.cores);
        first.release().unwrap();
        assert_eq!(held_cores(Some(&path)).unwrap(), second.cores);
        second.release().unwrap();
        let _ = fs::remove_file(&path);
        let _ = fs::remove_file(format!("{}.lock", path.display()));
    }
}
